src/python/pants/backend/python/macros/poetry_requirements.py

```diff
@@ -0,0 +1,287 @@
+        return ""
+    valid_specifiers = "<>!~= "
+    or_and_split = [[j.strip() for j in i.split(",")] for i in constr.split("||")]
+    ver_parsed = [[handle_str_attr("", j) for j in i] for i in or_and_split]
+
+    def conv_and(lst: list[str]) -> list:
+        return list(itertools.chain(*[i.split(",") for i in lst]))
+
+    def prepend(version: str) -> str:
+        return (
+            f"python_version{''.join(i for i in version if i in valid_specifiers)} '"
+            f"{''.join(i for i in version if i not in valid_specifiers)}'"
+        )
+
+    prepend_and_clean = [
+        [prepend(".".join(j.split(".")[:2])) for j in conv_and(i)] for i in ver_parsed
+    ]
+    return (
+        f"{'(' if len(or_and_split) > 1 else ''}"
+        f"{') or ('.join([' and '.join(i) for i in prepend_and_clean])}"
+        f"{')' if len(or_and_split) > 1 else ''}"
+    )
+
+
+def handle_dict_attr(proj_name: str, attributes: dict[str, str]) -> str:
+    def produce_match(sep: str, feat: Optional[str]) -> str:
+        return f"{sep}{feat}" if feat else ""
+
+    git_lookup = attributes.get("git")
+    if git_lookup is not None:
+        rev_lookup = produce_match("#", attributes.get("rev"))
+        branch_lookup = produce_match("@", attributes.get("branch"))
+        tag_lookup = produce_match("@", attributes.get("tag"))
+
+        return f"{proj_name} @ git+{git_lookup}{tag_lookup}{branch_lookup}{rev_lookup}"
+
+    version_lookup = attributes.get("version")
+    path_lookup = attributes.get("path")
+    if path_lookup is not None:
+        return f"{proj_name} @ file://{path_lookup}"
+    url_lookup = attributes.get("url")
+    if url_lookup is not None:
+        return f"{proj_name} @ {url_lookup}"
+    if version_lookup is not None:
+        markers_lookup = produce_match(";", attributes.get("markers"))
+        python_lookup = parse_python_constraint(attributes.get("python"))
+        version_parsed = handle_str_attr(proj_name, version_lookup)
+        return (
+            f"{version_parsed}"
+            f"{markers_lookup}"
+            f"{' and ' if python_lookup and markers_lookup else (';' if python_lookup else '')}"
+            f"{python_lookup}"
+        )
+    else:
+        raise AssertionError(
+            (
+                f"{proj_name} is not formatted correctly; at"
+                " minimum provide either a version, url, path or git location for"
+                " your dependency. "
+            )
+        )
+
+
+def parse_single_dependency(
+    proj_name: str, attributes: str | dict[str, Any] | list[dict[str, Any]]
+) -> tuple[Requirement, ...]:
+    if isinstance(attributes, str):
+        return (Requirement.parse(handle_str_attr(proj_name, attributes)),)
+    elif isinstance(attributes, dict):
+        return (Requirement.parse(handle_dict_attr(proj_name, attributes)),)
+    elif isinstance(attributes, list):
+        return tuple([Requirement.parse(handle_dict_attr(proj_name, attr)) for attr in attributes])
```

Comment:
You can remove the inner list comprehension and use a generator instead:

```suggestion
        return tuple(Requirement.parse(handle_dict_attr(proj_name, attr)) for attr in attributes)
```